docs/adrs/002-Retry_endpoint.md

```diff
@@ -0,0 +1,81 @@
+# Retry Endpoint
+
+Date: 2021-04-27
+
+The current requirement for the Batch/Bulk UI is to be able to retry jobs. Using a job id, it should be possible to get information from the database.
+
+The current implementation of retry is to run jobs using the `run_job` or `run_job_batch` endpoint. This is not adequate due to the following deficiencies:
+
+
+* Lack of book-keeping for the parent job and the child job relationship: Launching a new job will not update the relationship between those jobs
+* Lack of book-keeping for number of retries for a particular job / set of job inputs
+* Lack of ability to launch multiple jobs using the `run_job_batch` endpoint without creating a new parent job
+* Lack of ability to ensure that the proper catalog version /git commit of an app is used from the front end based on a tag, such as "beta/dev/release"
+* Lack of ability to specify which retries succeeded and which ones failed during submit time. 
+
+* Code is split more than is necessary
+
+## Author(s)
+
+@bio-boris
+
+## Status
+
+Pending
+
+## Alternatives Considered
+
+* Not book-keeping, or doing minimal book-keeping and calling run_job multiple times
+* Re-writing run_job/run_job_batch to address the aforementioned deficiencies
+* Creating a retry endpoint dedicated to addressing book-keeping and job launching features
+
+## Decision Outcome
+
+Pending
+
+* Link to spec file with inputs and outputs for the retry endpoint goes here
+* Link to Jira Ticket with business logic documentation for success and cancel cases
+
+### Possible additional things to think about
+* Creating sets of objects and what to do at the end of a batch run
+* What to do about a set if a child task fails during processing 
+* Convenience endpoints that operate on the parent_job_id or list of child job ids may be out of scope (e.g. cancel all jobs with a certain status)
+
```

Comment:
- should retries be limited to jobs with certain statuses? e.g. it would make little sense to retry a job that completed successfully or a job that is queued or running.
- should the number of retries of a job be limited, and if so, where? e.g. a `max_retries` field in the parent job? wait and see whether people attempt to rerun jobs that have already failed nine zillion times?

Reply:
- how to prevent incorrect parent-child relationships being created -- should the client be allowed to specify a parent ID? Is it currently possible to add a new child to a parent job if the child is a new job, rather than an existing job ID / set of params that is being rerun?

Reply:
* If you want to make sure the analysis is reproducible, you may want to retry that job, but then you would lose it in the UI, so you probably wouldn't want to do that. If you suspect the job is forever queued or forever stuck running, you may want to retry it. 
* Maybe have a "retry" field in the child job, and once it has reached a threshold, post an alert that says "Are you really sure you want to retry this job, it has failed "X" times" ?
* Any child can claim a parent right now currently, as per kbparallels. It is not currently possible to add a new child job to a parent job in any way.

Reply:
> If you suspect the job is forever queued or forever stuck running, you may want to retry it.
Is it possible that a job could get stuck in the queued or running states? Are there periodic checks done by ee2 to see if a job is still active?

Reply:
> Any child can claim a parent right now currently, as per kbparallels. It is not currently possible to add a new child job to a parent job in any way.

That seems contradictory -- if any child can claim a parent, you could just create a new job and give it whatever parent you want to add it to.

I would be in favour of preventing external modules from creating parent-child relationships where possible, to reduce the scope for crazy job structures like circular parent-child relationships or huge job trees.

Reply:
* No there are no checks done by ee2 to see if a job is still active. There are cases when the job gets stuck where that check wouldn't work either.

The parent job has a list of child jobs 
```
Parent
child_jobs = [1,2,3]
```

You can claim a parent, but the child_jobs array doesn't get updated. Just because you claim the job is your parent, doesnt mean the parent recognizes you as your child.

It seems to be in scope to support apps like metabat, which does exactly that, so we can't prevent that easily right now.

Reply:
Do you know if there are any other apps that do this? It might be easier to edit metabat and to ensure that ee2 is in charge of all job relationships than to have to work around a few exceptions.

Reply:
I'm with AJ on thinking it's scary that any process can claim any job as a parent job for a job that it submits.

For retries for us, I don't think that's necessary as the parent job can just be looked up from the job record created by `run_job_batch`.